GNX/man/man1/mail.1
.ig
	@(#)mail.1	1.3	6/29/83
	@(#)Copyright (C) 1983 by National Semiconductor Corp.
..
.TH MAIL 1
.UC 4
.SH NAME
mail \- send and receive mail
.SH SYNOPSIS
.B mail
[
.B \-f
[
name
]
]
[
people ...
]
.SH INTRODUCTION
.I Mail
is a intelligent mail processing system, which has
a command syntax reminiscent of
.I ed
with lines replaced by messages.
.PP
.I "Sending mail.\ "
To send a message to one or more other people,
.I mail
can be invoked with arguments which are the names of people to
send to.  You are then expected to type in your message, followed
by an \s-2EOT\s0 (control\-D) at the beginning of a line.
The section below, labeled
.I "Replying to or originating mail,"
describes some features of
.I mail
available to help you compose your letter.
.PP
.I "Reading mail.\ "
In normal usage,
.I mail
is given no arguments and checks your mail out of the
post office, then
printing out a one line header of each message there.
The current message is initially the first message (numbered 1)
and can be printed using the
.B print
command (which can be abbreviated \fBp\fR).
You can move among the messages much as you move between lines in
.I ed,
with the commands `+' and `\-' moving backwards and forwards, and
simple numbers typing the addressed message.
.PP
.I "Disposing of mail.\ "
After examining a message you can
.B delete
(\fBd\fR)
the message or
.B reply
(\fBr\fR)
to it.
Deletion causes the
.I mail
program to forget about the message.
This is not irreversible, the message can be
.B undeleted
(\fBu\fR)
by giving its number, or the
.I mail
session can be aborted by giving the
.B exit
(\fBx\fR)
command.
Deleted messages will, however, usually disappear never to be seen again.
.PP
.I "Specifying messages.\ "
Commands such as
.B print
and
.B delete
often can be given a list of message numbers as argument to apply
to a number of messages at once.
Thus ``delete 1 2'' deletes messages 1 and 2, while ``delete 1\-5''
deletes messages 1 through 5.
The special name ``*'' addresses all messages, and ``$'' addresses
the last message; thus the command
.B top
which prints the first few lines of a message could be used in
``top *'' to print the first few lines of all messages.
.PP
.I "Replying to or originating mail.\ "
You can use the
.B reply
command to
set up a response to a message, sending it back to the
person who it was from.
Text you then type in, up to an end-of-file (or a line consisting only
of a ``.'') defines the contents of
the message.
While you are composing a message,
.I mail
treats lines beginning with the character `~' specially.
For instance, typing ``~m'' (alone on a line) will place a copy
of the current message into the response right shifting it by a tabstop.
Other escapes will set up subject fields, add and delete recipients
to the message and allow you to escape to an editor to revise the
message or to a shell to run some commands.  (These options
will be given in the summary below.)
.PP
.I "Ending a mail processing session.\ "
You can end a
.I mail
session with the
.B quit
(\fBq\fR)
command.
Messages which have been examined go to your
.I mbox
file unless they have been deleted in which case they are discarded.
Unexamined messages go back to the post office.
The
.B \-f
option causes
.I mail
to read in the contents of your
.I mbox
(or the specified file)
for processing; when you
.B quit
.I mail
writes undeleted messages back to this file.
.PP
.I "Personal and systemwide distribution lists.\ "
It is also possible to create a personal distribution lists so that,
for instance, you can send mail to ``cohorts'' and have it go
to a group of people.
Such lists can be defined by placing a line like
.IP
alias cohorts bill ozalp sklower jkf mark cory:kridle
.PP
in the file \&.mailrc in your home directory.
The current list of such aliases can be displayed by the
.B alias
.B (a)
command in
.I mail.
System wide distribution lists can be created by editing
/usr/lib/aliases, see
.IR aliases (5)
and
.IR delivermail (8);
these are kept in a slightly different syntax.
In mail you send, personal aliases will be expanded in mail sent
to others so that they will be able to \fBreply\fR to the recipients.
System wide \fIaliases\fR are not expanded when the mail is sent,
but any reply returned to the machine will have the system wide
alias expanded as all mail goes through
.I delivermail.
If you edit
/usr/lib/aliases, you must run the program
.IR newaliases (1).
.PP
.I "Network mail (ARPA, UUCP, Berknet)\ "
Mail to sites on the ARPA network and sites within Bell laboratories
can be sent using ``name@site'' for ARPA-net sites or ``machine!user''
for Bell labs sites, provided appropriate gateways are known to the
system.  (Be sure to escape the ! in Bell sites when giving it on a
.I csh
command line by preceding it with an \e.
Machines on an instance of the Berkeley network are addressed as
``machine:user'', e.g. ``csvax:bill''.  When addressed from the
arpa-net, ``csvax:bill'' is known as ``csvax.bill@berkeley''.
.PP
.I Mail
has a number of options which can be
.B set
in the
.I \&.mailrc
file to alter its behavior; thus ``set askcc'' enables the ``askcc''
feature.  (These options are summarized below.)
.SH SUMMARY
(Adapted from the `Mail Reference Manual')
Each command is typed on a line by itself, and may take arguments
following the command word.  The command need not be typed in its
entirety \- the first command which matches the typed prefix is used.
For the commands which take message lists as arguments, if no message
list is given, then the next message forward which satisfies the
command's requirements is used.  If there are no messages forward of
the current message, the search proceeds backwards, and if there are no
good messages at all,
.I mail
types ``No applicable messages'' and
aborts the command.
.TP 12n
.B \-
Goes to the previous message and prints it out.  If given a numeric
argument
.I n ,
goes to the
.I n th
previous message and prints it.
.TP
.B ?
Prints a brief summary of commands.
.TP
.B !
Executes the \s-2UNIX\s0 shell command which follows.
.TP
.B alias
(\fBa\fR) With no arguments, prints out all currently-defined aliases.  With one
argument, prints out that alias.  With more than one argument, adds the
users named in the second and later arguments to the alias named in the
first argument.
.TP
.B chdir
(\fBc\fR) Changes the user's working directory to that specified, if given.  If
no directory is given, then changes to the user's login directory.
.TP
.B delete
(\fBd\fR) Takes a list of messages as argument and marks them all as deleted.
Deleted messages will not be saved in
.I mbox ,
nor will they be available for most other commands.  
.TP
.B dp
(also \fBdt\fR) Deletes the current message and prints the next message.
If there is no next message, 
.I mail
says ``at EOF.''
.TP
.B edit
(\fBe\fR) Takes a list of messages and points the text editor at each one in
turn.  On return from the editor, the message is read back in.
.TP
.B exit
(\fBex\fR or \fBx\fR) Effects an immediate return to the Shell without
modifying the user's system mailbox, his
.I mbox
file, or his edit file in
.B \-f .
.TP
.B from
(\fBf\fR) Takes a list of messages and prints their message headers.
.TP
.B headers
(\fBh\fR) Lists the current range of headers, which is an 18 message group.  If
a ``+'' argument is given, then the next 18 message group is printed, and if
a ``\-'' argument is given, the previous 18 message group is printed.
.TP
.B help
A synonym for ?
.TP
.B hold
(\fBho\fR, also \fBpreserve\fR) Takes a message list and marks each
message therein to be saved in the
user's system mailbox instead of in
.I mbox.
Does not override the
.B delete
command.
.TP
.B mail
(\fBm\fR) Takes as argument login names and distribution group names and sends
mail to those people.
.TP
.B next
(\fBn\fR like \fB+\fR or CR) Goes to the next message in sequence and types it.
With an argument list, types the next matching message.
.TP
.B preserve
A synonym for
.B hold.
.TP
.B print
(\fBp\fR) Takes a message list and types out each message on the user's
terminal.
.TP
.B quit
(\fBq\fR) Terminates the session, saving all undeleted, unsaved messages in
the user's
.I mbox
file in his login directory, preserving all messages marked with
.B hold
or
.B preserve
or never referenced
in his system mailbox, and removing all other messages from his system
mailbox.  If new mail has arrived during the session, the message
``You have new mail'' is given.  If given while editing a
mailbox file with the
.B \-f
flag, then the edit file is rewritten.  A return to the Shell is
effected, unless the rewrite of edit file fails, in which case the user
can escape with the
.B exit
command.
.TP
.B reply
(\fBr\fR) Takes a message list and sends mail to each message author just like
the
.B mail
command.  The default message must not be deleted.
.TP
.B respond
A synonym for
.B reply .
.TP
.B save
(\fBs\fR) Takes a message list and a filename and appends each message in
turn to the end of the file.  The filename in quotes, followed by the line
count and character count is echoed on the user's terminal.  
.TP
.B set
(\fBse\fR) With no arguments, prints all variable values.  Otherwise, sets
option.  Arguments are of the form
``option=value''
or
``option.''
.TP
.B shell
(\fBsh\fR) Invokes an interactive version of the shell.
.TP
.B size
Takes a message list and prints out the size in characters of each
message.
.TP
.B top
Takes a message list and prints the top few lines of each.  The number of
lines printed is controlled by the variable
.B toplines
and defaults to five.
.TP
.B type
(\fBt\fR) A synonym for
.B print .
.TP
.B unalias
Takes a list of names defined by
.B alias
commands and discards the remembered groups of users.  The group names
no longer have any significance.
.TP
.B undelete
(\fBu\fR) Takes a message list and marks each one as
.I not
being deleted.
.TP
.B unset
Takes a list of option names and discards their remembered values;
the inverse of
.B set .
.TP
.B visual
(\fBv\fR) Takes a message list and invokes the display editor on each message.
.TP
.B write
(\fBw\fR) A synonym for
.B save .
.TP
.B xit
(\fBx\fR) A synonym for
.B exit .
.PP
Here is a summary of the tilde escapes,
which are used when composing messages to perform
special functions.  Tilde escapes are only recognized at the beginning
of lines.  The name
``tilde\ escape''
is somewhat of a misnomer since the actual escape character can be set
by the option
.B escape.
.TP 12n
\fB~!\fRcommand
Execute the indicated shell command, then return to the message.
.TP
\fB~c\fR name ...
Add the given names to the list of carbon copy recipients.
.TP
.B ~d
Read the file ``dead.letter'' from your home directory into the message.
.TP
.B ~e
Invoke the text editor on the message collected so far.  After the
editing session is finished, you may continue appending text to the
message.
.TP
.B ~h
Edit the message header fields by typing each one in turn and allowing
the user to append text to the end or modify the field by using the
current terminal erase and kill characters.
.TP
\fB~m\fR messages
Read the named messages into the message being sent, shifted right one
tab.  If no messages are specified, read the current message.
.TP
.B ~p
Print out the message collected so far, prefaced by the message header
fields.
.TP
.B ~q
Abort the message being sent, copying the message to
``dead.letter''
in your home directory if
.B save
is set.
.TP
\fB~r\fR filename
Read the named file into the message.
.TP
\fB~s\fR string
Cause the named string to become the current subject field.
.TP
\fB~t\fR name ...
Add the given names to the direct recipient list.
.TP
.B ~v
Invoke an alternate editor (defined by the VISUAL option) on the
message collected so far.  Usually, the alternate editor will be a
screen editor.  After you quit the editor, you may resume appending
text to the end of your message.
.TP
\fB~w\fR filename
Write the message onto the named file.
.TP
\fB~\||\|\fRcommand
Pipe the message through the command as a filter.  If the command gives
no output or terminates abnormally, retain the original text of the
message.  The command
.IR fmt (1)
is often used as
.I command
to rejustify the message.
.TP
\fB~~\fRstring
Insert the string of text in the message prefaced by a single ~.  If
you have changed the escape character, then you should double
that character in order to send it.
.PP
Options are controlled via the
.B set
and
.B unset
commands.  Options may be either binary, in which case it is only
significant to see whether they are set or not, or string, in which
case the actual value is of interest.
The binary options include the following:
.TP 15n
.B append
Causes messages saved in
.I mbox
to be appended to the end rather than prepended.
(This is set in
/usr/lib/Mail.rc
on version 7 systems.)
.TP
.B ask
Causes
.I mail
to prompt you for the subject of each message you send.  If
you respond with simply a newline, no subject field will be sent.
.TP
.B askcc
Causes you to be prompted for additional carbon copy recipients at the
end of each message.  Responding with a newline indicates your
satisfaction with the current list.
.TP
.B autoprint
Causes the
.B delete
command to behave like
.B dp
\- thus, after deleting a message, the next one will be typed
automatically.
.TP
.B ignore
Causes interrupt signals from your terminal to be ignored and echoed as
@'s.
.TP
.B metoo
Usually, when a group is expanded that contains the sender, the sender
is removed from the expansion.  Setting this option causes the sender
to be included in the group.
.TP
.B quiet
Suppresses the printing of the version when first invoked.
.TP
.B save
Causes the message collected prior to a interrupt to be saved on
the file
``dead.letter''
in your home directory on receipt of two interrupts (or after a \fB~q\fR.)
.PP
The following options have string values:
.TP 15n
EDITOR
Pathname of the text editor to use in the
.B edit
command and ~e escape.  If not defined, then a default editor is used.
.TP
SHELL
Pathname of the shell to use in the
.B !
command and the ~! escape.  A default shell is used if this option is
not defined.
.TP

GNX/src/cmd/mail.c

/*
 * mail.c: version 1.2 of 6/6/83
 * Mesa Unix System Command Source File
 */
# ifdef SCCS
static char *sccsid = "@(#)mail.c	1.2 (NSC) 6/6/83";
# endif

#include <ctype.h>
#include <stdio.h>
#include <pwd.h>
#include <utmp.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <setjmp.h>
#include <sysexits.h>

/*
static char SccsId[] = "@(#)mail.c	4.2	5/8/81";
*/

#define DELIVERMAIL	"/etc/delivermail"


/*copylet flags */
	/*remote mail, add rmtmsg */
#define REMOTE	1
	/* zap header and trailing empty line */
#define ZAP	3
#define ORDINARY 2
#define	FORWARD	4
#define	LSIZE	256
#define	MAXLET	300	/* maximum number of letters */
#define	MAILMODE (~0644)		/* mode of created mail */
# ifndef DELIVERMAIL
#define	RMAIL	"/usr/net/bin/sendberkmail"
#define LOCNAM1	"csvax"
#define LOCNAM2	"ucbvax"
#define LOCNAM3	"vax"
#define LOCNAM4	"v"
# endif

char	line[LSIZE];
char	resp[LSIZE];
struct let {
	long	adr;
	char	change;
} let[MAXLET];
int	nlet	= 0;
char	lfil[50];
long	iop, time();
char	*getenv();
char	*index();
char	lettmp[] = "/tmp/maXXXXX";
char	maildir[] = "/usr/spool/mail/";
char	mailfile[] = "/usr/spool/mail/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
char	dead[] = "dead.letter";
char	sysnam[32] = "";
char	*thissys = sysnam;
char	*netname = "vax";
char	forwmsg[] = " forwarded\n";
FILE	*tmpf;
FILE	*malf;
char	*my_name;
char	*getlogin();
struct	passwd	*getpwuid();
int	error;
int	changed;
int	forward;
char	from[] = "From ";
long	ftell();
int	delete();
char	*ctime();
int	flgf;
int	flgp;
int	delflg = 1;
int	hseqno;
jmp_buf	sjbuf;
int	rmail;

main(argc, argv)
char **argv;
{
	register i;
	char sobuf[BUFSIZ];

	setbuf(stdout, sobuf);
	mktemp(lettmp);
	unlink(lettmp);
	my_name = getlogin();
	if (my_name == NULL || strlen(my_name) == 0) {
		struct passwd *pwent;
		pwent = getpwuid(getuid());
		if (pwent==NULL)
			my_name = "???";
		else
			my_name = pwent->pw_name;
	}
	if(setjmp(sjbuf)) done();
	for (i=0; i<20; i++)
		setsig(i, delete);
	tmpf = fopen(lettmp, "w");
	if (tmpf == NULL) {
		fprintf(stderr, "mail: cannot open %s for writing\n", lettmp);
		done();
	}
	if (argv[0][0] == 'r')
		rmail++;
	if (argv[0][0] != 'r' &&	/* no favors for rmail*/
	   (argc == 1 || argv[1][0] == '-' && !any(argv[1][1], "rhd")))
		printmail(argc, argv);
	else
		sendmail(argc, argv);
	done();
}

setsig(i, f)
int i;
int (*f)();
{
	if(signal(i, SIG_IGN)!=SIG_IGN)
		signal(i, f);
}

any(c, str)
	register int c;
	register char *str;
{

	while (*str)
		if (c == *str++)
			return(1);
	return(0);
}

printmail(argc, argv)
char **argv;
{
	int flg, i, j, print;
	char *p, *getarg();
	struct stat statb;

	setuid(getuid());
	cat(mailfile, maildir, my_name);
	if (stat(mailfile, &statb) >= 0
	    && (statb.st_mode & S_IFMT) == S_IFDIR) {
		strcat(mailfile, "/");
		strcat(mailfile, my_name);
	}
	for (; argc>1; argv++, argc--) {
		if (argv[1][0]=='-') {
			if (argv[1][1]=='q')
				delflg = 0;
			else if (argv[1][1]=='p') {
				flgp++;
				delflg = 0;
			} else if (argv[1][1]=='f') {
				if (argc>=3) {
					strcpy(mailfile, argv[2]);
					argv++;
					argc--;
				}
			} else if (argv[1][1]=='r') {
				forward = 1;
			} else if (argv[1][1]=='h') {
				forward = 1;
			} else {
				fprintf(stderr, "mail: unknown option %c\n", argv[1][1]);
				done();
			}
		} else
			break;
	}
	malf = fopen(mailfile, "r");
	if (malf == NULL) {
		fprintf(stdout, "No mail.\n");
		return;
	}
	lock(mailfile);
	copymt(malf, tmpf);
	fclose(malf);
	fclose(tmpf);
	unlock();
	tmpf = fopen(lettmp, "r");

	changed = 0;
	print = 1;
	for (i = 0; i < nlet; ) {
		j = forward ? i : nlet - i - 1;
		if(setjmp(sjbuf)) {
			print=0;
		} else {
			if (print)
				copylet(j, stdout, ORDINARY);
			print = 1;
		}
		if (flgp) {
			i++;
			continue;
		}
		setjmp(sjbuf);
		fprintf(stdout, "? ");
		fflush(stdout);
		if (fgets(resp, LSIZE, stdin) == NULL)
			break;
		switch (resp[0]) {

		default:
			fprintf(stderr, "usage\n");
		case '?':
			print = 0;
			fprintf(stderr, "q\tquit\n");
			fprintf(stderr, "x\texit without changing mail\n");
			fprintf(stderr, "p\tprint\n");
			fprintf(stderr, "s[file]\tsave (default mbox)\n");
			fprintf(stderr, "w[file]\tsame without header\n");
			fprintf(stderr, "-\tprint previous\n");
			fprintf(stderr, "d\tdelete\n");
			fprintf(stderr, "+\tnext (no delete)\n");
			fprintf(stderr, "m user\tmail to user\n");
			fprintf(stderr, "! cmd\texecute cmd\n");
			break;

		case '+':
		case 'n':
		case '\n':
			i++;
			break;
		case 'x':
			changed = 0;
		case 'q':
			goto donep;
		case 'p':
			break;
		case '^':
		case '-':
			if (--i < 0)
				i = 0;
			break;
		case 'y':
		case 'w':
		case 's':
			flg = 0;
			if (resp[1] != '\n' && resp[1] != ' ') {
				printf("illegal\n");
				flg++;
				print = 0;
				continue;
			}
			if (resp[1] == '\n' || resp[1] == '\0') {
				p = getenv("HOME");
				if(p != 0)
					cat(resp+1, p, "/mbox");
				else
					cat(resp+1, "", "mbox");
			}
			for (p = resp+1; (p = getarg(lfil, p)) != NULL; ) {
				malf = fopen(lfil, "a");
				if (malf == NULL) {
					fprintf(stdout, "mail: cannot append to %s\n", lfil);
					flg++;
					continue;
				}
				copylet(j, malf, resp[0]=='w'? ZAP: ORDINARY);
				fclose(malf);
			}
			if (flg)
				print = 0;
			else {
				let[j].change = 'd';
				changed++;
				i++;
			}
			break;
		case 'm':
			flg = 0;
			if (resp[1] == '\n' || resp[1] == '\0') {
				i++;
				continue;
			}
			if (resp[1] != ' ') {
				printf("invalid command\n");
				flg++;
				print = 0;
				continue;
			}
			for (p = resp+1; (p = getarg(lfil, p)) != NULL; )
				if (!sendrmt(j, lfil, "/bin/mail"))	/* couldn't send it */
					flg++;
			if (flg)
				print = 0;
			else {
				let[j].change = 'd';
				changed++;
				i++;
			}
			break;
		case '!':
			system(resp+1);
			printf("!\n");
			print = 0;
			break;
		case 'd':
			let[j].change = 'd';
			changed++;
			i++;
			if (resp[1] == 'q')
				goto donep;
			break;
		}
	}
   donep:
	if (changed)
		copyback();
}

copyback()	/* copy temp or whatever back to /usr/spool/mail */
{
	register i, n, c;
	int new = 0;
	struct stat stbuf;

	signal(SIGINT, SIG_IGN);
	signal(SIGHUP, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	lock(mailfile);
	stat(mailfile, &stbuf);
	if (stbuf.st_size != let[nlet].adr) {	/* new mail has arrived */
		malf = fopen(mailfile, "r");
		if (malf == NULL) {
			fprintf(stdout, "mail: can't re-read %s\n", mailfile);
			done();
		}
		fseek(malf, let[nlet].adr, 0);
		fclose(tmpf);
		tmpf = fopen(lettmp, "a");
		fseek(tmpf, let[nlet].adr, 0);
		while ((c = fgetc(malf)) != EOF)
			fputc(c, tmpf);
		fclose(malf);
		fclose(tmpf);
		tmpf = fopen(lettmp, "r");
		let[++nlet].adr = stbuf.st_size;
		new = 1;
	}
	malf = fopen(mailfile, "w");
	if (malf == NULL) {
		fprintf(stderr, "mail: can't rewrite %s\n", lfil);
		done();
	}
	n = 0;
	for (i = 0; i < nlet; i++)
		if (let[i].change != 'd') {
			copylet(i, malf, ORDINARY);
			n++;
		}
	fclose(malf);
	if (new)
		fprintf(stdout, "new mail arrived\n");
	unlock();
}

copymt(f1, f2)	/* copy mail (f1) to temp (f2) */
FILE *f1, *f2;
{
	long nextadr;

	nlet = nextadr = 0;
	let[0].adr = 0;
	while (fgets(line, LSIZE, f1) != NULL) {
		if (isfrom(line))
			let[nlet++].adr = nextadr;
		nextadr += strlen(line);
		fputs(line, f2);
	}
	let[nlet].adr = nextadr;	/* last plus 1 */
}

copylet(n, f, type) FILE *f;
{	int ch, k;
	int hnamelen = 32;
	int *namelen;
	namelen = (int *) &hnamelen;
	fseek(tmpf, let[n].adr, 0);
	k = let[n+1].adr - let[n].adr;
	while(k-- > 1 && (ch=fgetc(tmpf))!='\n')
		if(type!=ZAP) fputc(ch,f);
	if(type==REMOTE) {
		if (*thissys == '\0')
			gethostname(thissys, namelen);
		fprintf(f, " remote from %s\n", thissys);
	}
	else if (type==FORWARD)
		fprintf(f, forwmsg);
	else if(type==ORDINARY)
		fputc(ch,f);
	while(k-->1)
		fputc(ch=fgetc(tmpf), f);
	if(type!=ZAP || ch!= '\n')
		fputc(fgetc(tmpf), f);
}

isfrom(lp)
register char *lp;
{
	register char *p;

	for (p = from; *p; )
		if (*lp++ != *p++)
			return(0);
	return(1);
}

sendmail(argc, argv)
char **argv;
{
	char truename[100];
	int first;
	register char *cp;
	int gaver = 0;
# ifdef DELIVERMAIL
	char *newargv[1000];
	register char **ap;
	register char **vp;
	int dflag;

	dflag = 0;
	if (argc < 1)
		fprintf(stderr, "puke\n");
	for (vp = argv, ap = newargv + 1; (*ap = *vp++) != 0; ap++)
	{
		if (ap[0][0] == '-' && ap[0][1] == 'd')
			dflag++;
	}
	if (!dflag)
	{
		/* give it to delivermail, rah rah! */
		unlink(lettmp);
		ap = newargv+1;
		if (rmail)
			*ap-- = "-s";
		*ap = "-delivermail";
		execv(DELIVERMAIL, ap);
		perror(DELIVERMAIL);
		exit(EX_UNAVAILABLE);
	}
# endif DELIVERMAIL

	truename[0] = 0;
	line[0] = '\0';

	/*
	 * When we fall out of this, argv[1] should be first name,
	 * argc should be number of names + 1.
	 */

	while (argc > 1 && *argv[1] == '-') {
		cp = *++argv;
		argc--;
		switch (cp[1]) {
		case 'r':
			if (argc <= 0) {
				usage();
				done();
			}
			gaver++;
			strcpy(truename, argv[1]);
			fgets(line, LSIZE, stdin);
			if (strcmpn("From", line, 4) == 0)
				line[0] = '\0';
			argv++;
			argc--;
			break;

		case 'h':
			if (argc <= 0) {
				usage();
				done();
			}
			hseqno = atoi(argv[1]);
			argv++;
			argc--;
			break;

# ifdef DELIVERMAIL
		case 'd':
			break;
# endif DELIVERMAIL
		
		default:
			usage();
			done();
		}
	}
	if (argc <= 1) {
		usage();
		done();
	}
	if (gaver == 0)
		strcpy(truename, my_name);
	/*
	if (argc > 4 && strcmp(argv[1], "-r") == 0) {
		strcpy(truename, argv[2]);
		argc -= 2;
		argv += 2;
		fgets(line, LSIZE, stdin);
		if (strcmpn("From", line, 4) == 0)
			line[0] = '\0';
	} else
		strcpy(truename, my_name);
	*/
	time(&iop);
	fprintf(tmpf, "%s%s %s", from, truename, ctime(&iop));
	iop = ftell(tmpf);
	flgf = 1;
	for (first = 1;; first = 0) {
		if (first && line[0] == '\0' && fgets(line, LSIZE, stdin) == NULL)
			break;
		if (!first && fgets(line, LSIZE, stdin) == NULL)
			break;
		if (line[0] == '.' && line[1] == '\n' && isatty(fileno(stdin)))
			break;
		if (isfrom(line))
			fputs(">", tmpf);
		fputs(line, tmpf);
		flgf = 0;
	}
	fputs("\n", tmpf);
	nlet = 1;
	let[0].adr = 0;
	let[1].adr = ftell(tmpf);
	fclose(tmpf);
	if (flgf)
		return;
	tmpf = fopen(lettmp, "r");
	if (tmpf == NULL) {
		fprintf(stderr, "mail: cannot reopen %s for reading\n", lettmp);
		return;
	}
	while (--argc > 0)
		if (!send(0, *++argv, truename))
			error++;
	if (error) {
		setuid(getuid());
		malf = fopen(dead, "w");
		if (malf == NULL) {
			fprintf(stdout, "mail: cannot open %s\n", dead);
			fclose(tmpf);
			return;
		}
		copylet(0, malf, ZAP);
		fclose(malf);
		fprintf(stdout, "Mail saved in %s\n", dead);
	}
	fclose(tmpf);
}

sendrmt(n, name, rcmd)
char *name;
char *rcmd;
{
	FILE *rmf, *popen();
	register char *p;
	char rsys[64], cmd[64];
	register local, pid;
	int sts;

	local = 0;
	if (index(name, '^')) {
		while (p = index(name, '^'))
			*p = '!';
		if (strncmp(name, "researc", 7)) {
			strcpy(rsys, "research");
			if (*name != '!')
				--name;
			goto skip;
		}
	}
	if (*name=='!')
		name++;
	for(p=rsys; *name!='!'; *p++ = *name++)
		if (*name=='\0') {
			local++;
			break;
		}
	*p = '\0';
	if ((!local && *name=='\0') || (local && *rsys=='\0')) {
		fprintf(stdout, "null name\n");
		return(0);
	}
skip:
	if ((pid = fork()) == -1) {
		fprintf(stderr, "mail: can't create proc for remote\n");
		return(0);
	}
	if (pid) {
		while (wait(&sts) != pid) {
			if (wait(&sts)==-1)
				return(0);
		}
		return(!sts);
	}
	setuid(getuid());
	if (local)
		sprintf(cmd, "%s %s", rcmd, rsys);
	else {
		if (index(name+1, '!'))
			sprintf(cmd, "uux - %s!rmail \\(%s\\)", rsys, name+1);
		else
			sprintf(cmd, "uux - %s!rmail %s", rsys, name+1);
	}
	if ((rmf=popen(cmd, "w")) == NULL)
		exit(1);
	copylet(n, rmf, local ? !strcmp(rcmd, "/bin/mail") ? FORWARD : ORDINARY : REMOTE);
	pclose(rmf);
	exit(0);
}

# ifndef DELIVERMAIL
/*
 * Send mail on the Berkeley network.
 * Sorry Bill, sendrmt() is so awful we just gave up.
 */

sendberkmail(n, name, fromaddr)
	char name[];
	char fromaddr[];
{
	char cmd[200];
	register FILE *cmdf;

	sprintf(cmd, "%s -h %d -f %s -t %s", RMAIL, hseqno, fromaddr, name);
	if ((cmdf = popen(cmd, "w")) == NULL) {
		perror(RMAIL);
		return(0);
	}
	copylet(n, cmdf, ORDINARY);
	pclose(cmdf);
	return(9);
}
# endif

usage()
{

	fprintf(stderr, "Usage: mail [ -f ] people . . .\n");
}

send(n, name, fromaddr)
int n;
char *name;
char *fromaddr;
{
	char file[100];
	register char *p;
	register mask;
	struct passwd *pw, *getpwnam();
	struct stat statb;
	char buf[128];
	int f;

# ifndef DELIVERMAIL
	stripfx(LOCNAM1, &name);
	stripfx(LOCNAM2, &name);
	stripfx(LOCNAM3, &name);
	stripfx(LOCNAM4, &name);
	if(*name == ':')name++;		/* skip colon in to-name */
	for(p=name; *p!=':' && *p!='!' && *p!='^' &&*p!='\0'; p++);
	/* if(*p == ':') return(sendrmt(n, name, RMAIL)); */
	if (*p == ':')
		return(sendberkmail(n, name, fromaddr));
	else if (*p=='\0' && strcmp(name, "msgs") == 0)
		return(sendrmt(n, "-s", "/usr/ucb/msgs"));
# endif
	for(p=name; *p!='!'&&*p!='^' &&*p!='\0'; p++)
		;
	if (*p == '!'|| *p=='^')
		return(sendrmt(n, name, 0));
	if ((pw = getpwnam(name)) == NULL) {
		fprintf(stdout, "mail: can't send to %s\n", name);
		return(0);
	}
	cat(file, maildir, name);
	if (stat(file, &statb) >= 0 && (statb.st_mode & S_IFMT) == S_IFDIR) {
		strcat(file, "/");
		strcat(file, name);
	}
	mask = umask(MAILMODE);
	malf = fopen(file, "a");
	umask(mask);
	if (malf == NULL) {
		fprintf(stdout, "mail: cannot append to %s\n", file);
		return(0);
	}
	lock(file);
	chown(file, pw->pw_uid, pw->pw_gid);
	{
		f = open("/dev/mail", 1);
		sprintf(buf, "%s@%d\n", name, ftell(malf)); 
	}
	copylet(n, malf, ORDINARY);
	if (f >= 0) {
		write(f, buf, strlen(buf)+1);
		close(f);
	}
	fclose(malf);
	unlock();
	return(1);
}

delete(i)
{
	setsig(i, delete);
	fprintf(stderr, "\n");
	if(delflg)
		longjmp(sjbuf, 1);
	done();
}

/*
 * Lock the specified mail file by setting the file mailfile.lock.
 * We must, of course, be careful to unlink the lock file by a call
 * to unlock before we stop.  The algorithm used here is to see if
 * the lock exists, and if it does, to check its modify time.  If it
 * is older than 30 seconds, we assume error and set our own file.
 * Otherwise, we wait for 5 seconds and try again.
 */

char	*maillock	= ".lock";		/* Lock suffix for mailname */
char	*lockname	= "/usr/spool/mail/tmXXXXXX";
char	locktmp[30];				/* Usable lock temporary */
char	curlock[50];				/* Last used name of lock */
int	locked;					/* To note that we locked it */

lock(file)
char *file;
{
	register int f;
	struct stat sbuf;
	long curtime;
	int statfailed;

	if (locked || flgf)
		return(0);
	strcpy(curlock, file);
	strcat(curlock, maillock);
	strcpy(locktmp, lockname);
	mktemp(locktmp);
	unlink(locktmp);
	statfailed = 0;
	for (;;) {
		f = lock1(locktmp, curlock);
		if (f == 0) {
			locked = 1;
			return(0);
		}
		if (stat(curlock, &sbuf) < 0) {
			if (statfailed++ > 5)
				return(-1);
			sleep(5);
			continue;
		}
		statfailed = 0;
		time(&curtime);
		if (curtime < sbuf.st_ctime + 30) {
			sleep(5);
			continue;
		}
		unlink(curlock);
	}
}

/*
 * Remove the mail lock, and note that we no longer
 * have it locked.
 */

unlock()
{

	unlink(curlock);
	locked = 0;
}

/*
 * Attempt to set the lock by creating the temporary file,
 * then doing a link/unlink.  If it fails, return -1 else 0
 */

lock1(tempfile, name)
	char tempfile[], name[];
{
	register int fd;

	fd = creat(tempfile, 0);
	if (fd < 0)
		return(-1);
	close(fd);
	if (link(tempfile, name) < 0) {
		unlink(tempfile);
		return(-1);
	}
	unlink(tempfile);
	return(0);
}

done()
{
	if(locked)
		unlock();
	unlink(lettmp);
	unlink(locktmp);
	exit(error);
}

cat(to, from1, from2)
char *to, *from1, *from2;
{
	int i, j;

	j = 0;
	for (i=0; from1[i]; i++)
		to[j++] = from1[i];
	for (i=0; from2[i]; i++)
		to[j++] = from2[i];
	to[j] = 0;
}

char *getarg(s, p)	/* copy p... into s, update p */
register char *s, *p;
{
	while (*p == ' ' || *p == '\t')
		p++;
	if (*p == '\n' || *p == '\0')
		return(NULL);
	while (*p != ' ' && *p != '\t' && *p != '\n' && *p != '\0')
		*s++ = *p++;
	*s = '\0';
	return(p);
}
# ifndef DELIVERMAIL
/*
	stripfx(prefix string, pointer to string)

	takes a ptr to string and compares it to prefix string.
	may be called multiple times
*/
stripfx(pfx, name)
	char *pfx;
	char **name;
{
	register char *cp = *name;

	while (*pfx && (*cp == *pfx || *cp == toupper(*pfx)))
		cp++, pfx++;
	if (*cp != ':' || *pfx != 0)
		return;
	*name = cp;
}
# endif
